version/main.go

```diff
@@ -12,65 +17,102 @@ import (
 	"strings"
 	"time"
 
-	"github.com/dagger/dagger/.dagger/internal/dagger"
 	"golang.org/x/mod/semver"
 )
 
-// VersionInfo is a helper for passing version information around.
-//
-// In general, it attempts to follow go's psedudoversioning:
-// https://go.dev/doc/modules/version-numbers
+func New(ctx context.Context,
+	// A directory containing all the inputs of the artifact to be versioned: the Dagger engine + CLI
+	// An input is any file that changes the artifact if it changes.
+	// This directory is used to compute a digest. If any input changes, the digest changes.
+	// - To avoid false positives, only include actual inputs
+	// - To avoid false negatives, include *all* inputs
+	// +optional
+	// +defaultPath="/"
+	// +ignore=[".git", "bin", "**/node_modules", "**/testdata", "**.changes", "docs", "helm", "release", "version", "modules", "*.md", "LICENSE", "NOTICE", "hack"]
+	inputs *dagger.Directory,
```

Comment:
Fun thought - what's weird is that we almost don't *want* a user to be able to override this contextual directory.

I wonder if we should have another helper here, something like `// +contextual` - any contextual arg *must* come from the context, and can't be overridden (this potentially helps solve my issue above).

Reply:
cc @vito @sipsma we bent over backwards in the context directory design to protect the argument system - make sure everything is an argument no matter what. I'm flagging this suggestion because if many users ask for the same thing, it may be a sign that we got it wrong?

Reply:
I'm not seeing why we wouldn't want this as an argument. 🤔 Especially after #8436 which would mean the `+ignore` rule is respected for non-contextual args too. Maybe that's the difference-maker?

Reply:
@vito the fear in this particular scenario, is that different callers might pass different input directories (because they can), resulting in different version strings within the same build, which might break things in horrible ways - which @jedevc would then have to debug :)